File: examples/examples.py
'''
2025-10-05

Examples of running imfcharts.

Data files are created by get_data_turkey_page1.py
and files are stored in data/turkey_page1/.
'''

import os
import sys
import glob

import numpy as np
import pandas as pd

from PyPDF2 import PdfWriter

import imfcharts
import imfcharts.charts
import imfcharts.mpl
import importlib
importlib.reload(imfcharts)
importlib.reload(imfcharts.charts)
importlib.reload(imfcharts.mpl)
from imfcharts import *

# Turn on interactive mode
# plt.ion()
# Clear all existing figures
plt.close('all')

outdir = 'pdf'

# ---------------------------------------------------------------------------------------------------------
# p. 1 Chart 1
# ---------------------------------------------------------------------------------------------------------
# Read in data
infilename = 'data/turkey_page1/fig1_chart1_gdp.csv'
if not os.path.isfile(infilename):
    print('File ' + infilename + ' does not exist')
    sys.exit()
df = pd.read_csv(infilename, index_col=0, parse_dates=[0])
cols = df.columns
linecol = cols[0]
barcols = cols[1:]

# Create Chart object
chart1 = Chart(df, linecols=linecol, barcols=barcols,
               dict_attrs={linecol : {'linewidth' : 4,
                                      # 'linestyle' : '--',
                                      'marker' : 'o',
                                      'markersize' : 10,
                                      'markerfacecolor' : 'white',
                                      'markeredgewidth' : 4,
                                      'markeredgecolor' : IMFBLUE}},
               title='Contributions to Real GDP Growth',
               subtitle='(Percent, q/q)',
               xrange='2023Q1:', yrange=[-3, 8],
               ncol_legend=2)
               # debug=True)

# Save
chart1.save(outdir + '/page1_chart1_gdp.pdf')

# ---------------------------------------------------------------------------------------------------------
# p. 1 Chart 2
# ---------------------------------------------------------------------------------------------------------
# Read in data
infilename = 'data/turkey_page1/fig1_chart2_labor.csv'
if not os.path.isfile(infilename):
    print('File ' + infilename + ' does not exist')
    sys.exit()
df = pd.read_csv(infilename, index_col=0, parse_dates=[0])

# Create Chart object
chart2 = Chart(df, linecols=df.columns[0], rlinecols=df.columns[1],
               title='Labor Utilization',
               subtitle='(Percent)',
               xrange='2023-01:2025-08', # yrange=[8, 11], ryrange=[10, 25],
               yrange=[8, 11],
               )
               # debug=True)

# Save
chart2.save(outdir + '/page1_chart2_labor.pdf')

# ---------------------------------------------------------------------------------------------------------
# p. 1 Chart 3
# ---------------------------------------------------------------------------------------------------------
# Read in data
infilename = 'data/turkey_page1/fig1_chart3_ip.csv'
if not os.path.isfile(infilename):
    print('File ' + infilename + ' does not exist')
    sys.exit()
df = pd.read_csv(infilename, index_col=0, parse_dates=[0])
df = df[['Industrial Production', 'Retail sales']]

# Create Chart object
chart3 = Chart(df, linecols=df.columns,
               title='Production Indicators',
               subtitle='(Percent, yoy)',
               legend_left=0.60, legend_width=0.40,
               xrange='2023-01:', yrange=[-10, 40],
               )
               # debug=True)

# Save
chart3.save(outdir + '/page1_chart3_ip.pdf')

# ---------------------------------------------------------------------------------------------------------
# p. 1 Chart 4
# ---------------------------------------------------------------------------------------------------------
# Read in data
infilename = 'data/turkey_page1/fig1_chart4_confidence.csv'
if not os.path.isfile(infilename):
    print('File ' + infilename + ' does not exist')
    sys.exit()
df = pd.read_csv(infilename, index_col=0, parse_dates=[0])

# Create Chart object
chart4 = Chart(df, linecols=df.columns,
               title='Economic Confidence Index',
               subtitle='(Seasonally adjusted, 100+=optimistic)',
               ncol_legend=2,
               xrange='2023-01:', yrange=[60, 140],
               ) # debug=True)

# Save
chart4.save(outdir + '/page1_chart4_confidence.pdf')

# ---------------------------------------------------------------------------------------------------------
# p. 1 Chart 5
# ---------------------------------------------------------------------------------------------------------
# Read in data
infilename = 'data/turkey_page1/fig1_chart5_partners.csv'
if not os.path.isfile(infilename):
    print('File ' + infilename + ' does not exist')
    sys.exit()
df = pd.read_csv(infilename, index_col=0, parse_dates=[0])

# Create Chart object
chart5 = Chart(df, linecols=df.columns,
               title='Major Trading Partner GDP Growth',
               subtitle='(Percent q/q, SAAR)',
               ncol_legend=2,
               legend_left=0.70, legend_width=0.30,
               xrange='2023-01:', yrange=[-1, 5],
               ) # debug=True)

# Save
chart5.save(outdir + '/page1_chart5_partners.pdf')

# ---------------------------------------------------------------------------------------------------------
# p. 1 Chart 6
# ---------------------------------------------------------------------------------------------------------
# Read in data
infilename = 'data/turkey_page1/fig1_chart6_trade.csv'
if not os.path.isfile(infilename):
    print('File ' + infilename + ' does not exist')
    sys.exit()
df = pd.read_csv(infilename, index_col=0, parse_dates=[0])

# Create Chart object
chart6 = Chart(df, linecols=df.columns,
               title='Major Trading Partner GDP Growth',
               subtitle='(Percent q/q, SAAR)',
               xrange='2019Q1:', yrange=[20, 55],
               ) # debug=True)

# Save
chart6.save(outdir + '/page1_chart6_trade.pdf')

# ---------------------------------------------------------------------------------------------------------
# Additional commands
# ---------------------------------------------------------------------------------------------------------
# Access figure object
fig = chart2.fig

# Show figure
# chart2.show()

# Add a horizontal line
chart2.add_hline(3)

# Apply style
# chart.apply(style)

# Show all charts
# plt.show()

# Delete previous merged file
mergedfilename = outdir + '/all.pdf'
if os.path.isfile(mergedfilename):
    os.remove(mergedfilename)

# Merge all files
merger = PdfWriter()
filenames = sorted(glob.glob(outdir + '/*.pdf'))
for filename in filenames:
    merger.append(filename)
merger.write(mergedfilename)
print('Created file ' + mergedfilename)

File: examples/examples.R
library(reticulate)
library(readr)
library(dplyr)       
library(tibble)     

# ---------------------------------------------------------------------
# import python modules
# ---------------------------------------------------------------------
imfcharts <- import("imfcharts")

# Attach all chart objects to the global environment
# This is the equivalent of doing
# from imfcharts import *
# in Python
for (name in names(imfcharts)) {
  assign(name, imfcharts[[name]], envir = .GlobalEnv)
}

pd <- import("pandas")
# matplotlib = import("matplotlib")
# matplotlib$use('Qt5Agg')

# ---------------------------------------------------------------------
# read & wrangle in R
# ---------------------------------------------------------------------
df_r <- read_csv("data/turkey_page6/fig6_chart1_current_account.csv") %>% 
  mutate(dates = lubridate::ymd(dates)) %>% 
  arrange(dates) %>% 
column_to_rownames("dates")

# Without conversion to index for dates
df_r2 <- read_csv("data/turkey_page6/fig6_chart1_current_account.csv") %>% 
  mutate(dates = lubridate::ymd(dates)) %>% 
  arrange(dates)

# ---------------------------------------------------------------------
# convert → pandas and make sure index is DatetimeIndex   
# ---------------------------------------------------------------------
py_df <- r_to_py(df_r)
py_df$index <- pd$to_datetime(py_df$index)

# ---------------------------------------------------------------------
# build dict_attrs ‒ This creates python dictionary in R
# ---------------------------------------------------------------------
attrs <- dict(
  `Risk weighted Assets (RHS)` = dict(color = IMFBLACK),
  Capital                       = dict(color = IMFBLUE),
  `Core capital`                = dict(color = IMFRED)
)

# ---------------------------------------------------------------------
# build the chart – NOTE: pass py_df, not df  and use r_to_py() to pass 
# python list to Chart()
# ---------------------------------------------------------------------
chart1 <- Chart(
  df_r,
  linecols  =  c("Capital", "Core capital"),
  rlinecols = "Risk weighted Assets (RHS)",
  attrs = attrs,
  title      = "Bank Capital",
  subtitle   = "(Percent of risk-weighted assets; RWA y/y percent change)",
  legend_left = 0.4,
  xrange     = "2022-01:",
  yrange     = c(10, 25),
  ryrange    = c(120, 200)
)
# chart1$show()
chart1$save("pdf/page6_chart1_bank_capital1_1.pdf")

chart2 <- Chart(
  df_r2,
  indexcol='dates',
  linecols  =  c("Capital", "Core capital"),
  rlinecols = "Risk weighted Assets (RHS)",
  attrs = attrs,
  title      = "Bank Capital",
  subtitle   = "(Percent of risk-weighted assets; RWA y/y percent change)",
  legend_left = 0.4,
  xrange     = "2022-01:",
  yrange     = c(10, 25),
  ryrange    = c(120, 200)
)
# chart2$show()
chart2$save("pdf/page6_chart1_bank_capital1_2.pdf")


chart3 <- Chart(
  py_df,
  linecols  =  c("Capital", "Core capital"),
  rlinecols = "Risk weighted Assets (RHS)",
  attrs = attrs,
  title      = "Bank Capital",
  subtitle   = "(Percent of risk-weighted assets; RWA y/y percent change)",
  legend_left = 0.4,
  xrange     = "2022-01:",
  yrange     = c(10, 25),
  ryrange    = c(120, 200)
)
# chart3$show()
chart3$save("pdf/page6_chart1_bank_capital1_3.pdf")
